File: agentic_os/mission/plan_select.py
"""P9a — Active plan selection: simulation as a first-class planning component.

    Candidate plans → policy pruning → simulation → cost/risk/success scoring → plan selection

Exploration is **bounded on purpose** — max candidate count, admissible (policy-scoped) providers
only, a one-swap neighbourhood instead of the full cross-product, and a switch margin the winner
must beat before the runtime abandons the default plan. Otherwise candidate generation blows up
combinatorially and non-determinism creeps back in.

The registry handed in is already policy-scoped, so every candidate is admissible by construction;
a binding the compiler still rejects (fail-closed) is simply dropped from the candidate set.
"""
from __future__ import annotations

from dataclasses import dataclass

from . import cost as costmod
from .compiler import compile_intent, CompileError
from .context import LocalContextRuntime
from .registry import CapabilityRegistry
from .scheduler import SchedulePolicy
from .simulator import simulate
from .types import ExecutionIntent, ExecutionPlan, Mission, SimResult


@dataclass
class PlanCandidate:
    label: str
    prefer: dict           # binding overrides {outcome: capability_name} ({} = the default plan)
    plan: ExecutionPlan
    projection: SimResult
    score: float
    is_default: bool = False
    selected: bool = False


def score_projection(proj: SimResult) -> float:
    """Cost/risk/success into one comparable number (higher = better). Over-budget is inadmissible."""
    if not proj.within_budget:
        return float("-inf")
    return proj.expected_success * 10.0 - proj.expected_usd - 0.5 * proj.expected_approvals


class ActivePlanner:
    def __init__(self, registry: CapabilityRegistry, policy: SchedulePolicy, learning=None, *,
                 max_candidates: int = 6, switch_margin: float = 0.05):
        self.registry = registry
        self.context = LocalContextRuntime(registry)   # binding resolves through the Context Runtime
        self.policy = policy
        self.learning = learning
        self.max_candidates = max(1, max_candidates)
        self.switch_margin = switch_margin

    def select(self, mission: Mission, intent: ExecutionIntent, *, revision: int = 1,
               reason: str = "initial") -> tuple[ExecutionPlan, list[PlanCandidate]]:
        cands: list[PlanCandidate] = []
        for label, pref in self._candidate_prefers(intent):
            try:
                plan = compile_intent(mission, intent, self.context,
                                      revision=revision, reason=reason, prefer=pref)
            except CompileError:
                continue                              # policy pruning: an inadmissible binding is dropped
            proj = simulate(plan, mission, self.registry, self.policy)
            cands.append(PlanCandidate(label=label, prefer=pref, plan=plan, projection=proj,
                                       score=score_projection(proj), is_default=(pref == {})))
        if not cands:
            # no admissible candidate — recompile the default so the caller gets the real CompileError
            plan = compile_intent(mission, intent, self.context, revision=revision, reason=reason)
            proj = simulate(plan, mission, self.registry, self.policy)
            cands = [PlanCandidate("default", {}, plan, proj, score_projection(proj), is_default=True)]

        default = next((c for c in cands if c.is_default), cands[0])
        best = max(cands, key=lambda c: c.score)
        # bounded: only abandon the default when a candidate clearly wins (confidence threshold)
        chosen = best if best.score > default.score + self.switch_margin else default
        chosen.selected = True
        chosen.plan.projection = chosen.projection
        return chosen.plan, cands

    def _candidate_prefers(self, intent: ExecutionIntent):
        """The default plan first, then a bounded one-swap neighbourhood: each alternative swaps a
        single outcome to a non-default admissible provider. One-swap, not the cross-product."""
        prefers: list[tuple[str, dict]] = [("default", {})]
        for step in intent.steps:
            ranked = costmod.rank_candidates(self.registry.providing(step.outcome))
            for alt in ranked[1:]:                    # skip the default (rank-best) provider
                prefers.append((f"{step.outcome}={alt.name}", {step.outcome: alt.name}))
                if len(prefers) >= self.max_candidates:
                    return prefers
        return prefers

File: go/mission/plan_select.go
package mission

// P9a — Active plan selection: simulation as a first-class planning component. Candidate plans →
// policy pruning → simulation → cost/risk/success scoring → selection, bounded (max candidates,
// admissible providers, one-swap neighbourhood, a switch margin protecting the default). Go port of
// agentic_os/mission/plan_select.py.

import "math"

type PlanCandidate struct {
	Label      string
	Prefer     map[string]string
	Plan       *ExecutionPlan
	Projection SimResult
	Score      float64
	IsDefault  bool
	Selected   bool
}

// scoreProjection folds cost/risk/success into one comparable number (higher = better). Over-budget
// is inadmissible.
func scoreProjection(proj SimResult) float64 {
	if !proj.WithinBudget {
		return math.Inf(-1)
	}
	return proj.ExpectedSuccess*10.0 - proj.ExpectedUSD - 0.5*float64(proj.ExpectedApprovals)
}

type ActivePlanner struct {
	registry      Registry
	policy        SchedulePolicy
	learning      *LearningRouter
	MaxCandidates int
	SwitchMargin  float64
}

func NewActivePlanner(registry Registry, policy SchedulePolicy, learning *LearningRouter) *ActivePlanner {
	return &ActivePlanner{registry: registry, policy: policy, learning: learning, MaxCandidates: 6, SwitchMargin: 0.05}
}

// Select generates candidates, prunes/simulates/scores, and picks — keeping the default unless a
// candidate clearly wins. Returns a CompileError only if even the default plan can't compile.
func (a *ActivePlanner) Select(mission *Mission, intent ExecutionIntent, revision int, reason string) (*ExecutionPlan, []*PlanCandidate, error) {
	var cands []*PlanCandidate
	for _, pp := range a.candidatePrefers(intent) {
		plan, err := CompileIntent(mission, intent, a.registry, revision, reason, pp.prefer)
		if err != nil {
			continue // policy pruning: an inadmissible binding is dropped
		}
		proj := Simulate(plan, mission, a.registry, &a.policy)
		cands = append(cands, &PlanCandidate{Label: pp.label, Prefer: pp.prefer, Plan: plan,
			Projection: proj, Score: scoreProjection(proj), IsDefault: len(pp.prefer) == 0})
	}
	if len(cands) == 0 {
		plan, err := CompileIntent(mission, intent, a.registry, revision, reason, nil)
		if err != nil {
			return nil, nil, err // surface the real CompileError to the caller (fail closed)
		}
		proj := Simulate(plan, mission, a.registry, &a.policy)
		cands = []*PlanCandidate{{Label: "default", Prefer: map[string]string{}, Plan: plan,
			Projection: proj, Score: scoreProjection(proj), IsDefault: true}}
	}

	def := cands[0]
	for _, c := range cands {
		if c.IsDefault {
			def = c
			break
		}
	}
	best := cands[0]
	for _, c := range cands {
		if c.Score > best.Score {
			best = c
		}
	}
	chosen := def
	if best.Score > def.Score+a.SwitchMargin {
		chosen = best
	}
	chosen.Selected = true
	chosen.Plan.Projection = &chosen.Projection
	return chosen.Plan, cands, nil
}

type preferEntry struct {
	label  string
	prefer map[string]string
}

// candidatePrefers: the default plan first, then a bounded one-swap neighbourhood (each alternative
// swaps a single outcome to a non-default admissible provider).
func (a *ActivePlanner) candidatePrefers(intent ExecutionIntent) []preferEntry {
	prefers := []preferEntry{{"default", map[string]string{}}}
	for _, step := range intent.Steps {
		ranked := rankCandidates(a.registry.Providing(step.Outcome))
		for i := 1; i < len(ranked); i++ { // skip the rank-best default provider
			alt := ranked[i]
			prefers = append(prefers, preferEntry{step.Outcome + "=" + alt.Name, map[string]string{step.Outcome: alt.Name}})
			if len(prefers) >= a.MaxCandidates {
				return prefers
			}
		}
	}
	return prefers
}
